src/server/tran_server.cpp

```diff
@@ -469,16 +470,22 @@ tran_server::connection_handler::~connection_handler ()
 void
 tran_server::connection_handler::disconnect_async (bool with_disc_msg)
 {
-  auto lockg = std::lock_guard <std::mutex> { m_disconn_mtx };
+  auto lockg = std::lock_guard <std::shared_mutex> { m_state_mtx };
 
-  if (m_disconn_future.valid ())
+  assert (m_state != state::CONNECTING);
```

Comment:
Q. is it possible to call this function with the state value as `state::CONNECTING`? 
It seems possible that disconnecting the TS-PS connection which is being established. 
(For instance, A situation where TS is started and immediately stopped.)

If so, separate handling might be required specifically for `state::CONNECTING`.
For example, it might be necessary to wait until state::CONNECTED is reached.

Reply:
It seems impossible because the OPEN->CONNECTING->CONNECTED transition happens at once with m_state_mtx acquired. Do you happen to see a room for that to happen?

Reply:
see `set_connection`

Reply:
Ah, I understand. I got a little confused about shared_mutex

Reply:
@joohok On second thought, it'd be better to deal with `state::CONNECTING` here for the near future. It's worth it when the catch-up process comes in and the transition, OPEN->CONNECTING->CONNECTED, doesn't happen at once. Thank you.